core/src/library.rs

```diff
@@ -579,6 +583,60 @@ impl<'gc> Library<'gc> {
         self.device_fonts.find(&query)
     }
 
+    fn sort_device_fonts(
+        &mut self,
+        query: &FontQuery,
+        ui: &dyn UiBackend,
+        renderer: &mut dyn RenderBackend,
+        gc_context: &Mutation<'gc>,
+    ) -> Vec<Font<'gc>> {
+        // First, ask the backend to sort the fonts for us.
+        let fonts = ui.sort_device_fonts(query, &mut |definition| {
+            self.register_device_font(gc_context, renderer, definition)
+        });
+
+        let fonts: Vec<Font<'gc>> = fonts
+            .iter()
+            .filter_map(|font_query| self.device_fonts.get(font_query))
+            .copied()
+            .collect();
+
+        if !fonts.is_empty() {
+            return fonts;
+        }
+
+        // When the backend failed (or doesn't support sorting fonts), fall back
+        // to loading one font only without sorting.
+        let font = self.get_or_load_device_font(
+            &query.name,
+            query.is_bold,
+            query.is_italic,
+            ui,
+            renderer,
+            gc_context,
+        );
+        font.map(|font| vec![font]).unwrap_or_default()
+    }
+
+    pub fn get_or_sort_device_fonts(
+        &mut self,
+        name: &str,
+        is_bold: bool,
+        is_italic: bool,
+        ui: &dyn UiBackend,
+        renderer: &mut dyn RenderBackend,
+        gc_context: &Mutation<'gc>,
+    ) -> Vec<Font<'gc>> {
+        let query = FontQuery::new(FontType::Device, name.to_owned(), is_bold, is_italic);
+        if let Some(fonts) = self.font_sort_cache.get(&query) {
+            return fonts.clone();
+        }
+
+        let fonts = self.sort_device_fonts(&query, ui, renderer, gc_context);
+        self.font_sort_cache.insert(query, fonts.clone());
```

Comment:
not that it's important for perf, but: couldn't this avoid a clone if the function returned a borrow? It should be possible, right? The only user in later commits just borrows the `.first()` anyway.

Reply:
Added a comment